daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala

```diff
@@ -378,14 +392,47 @@ object Value extends CidContainer1WithDefaultCidResolver[Value] {
       implicit val `V0 Order`: Order[V0] = Order.fromScalaOrdering[String] contramap (_.coid)
     }
 
-    final case class V1(discriminator: crypto.Hash, suffix: Bytes = Bytes.Empty)
-        extends AbsoluteContractId {
+    final class V1(val discriminator: crypto.Hash, val suffix: Bytes) extends AbsoluteContractId {
       lazy val toBytes: Bytes = V1.prefix ++ discriminator.bytes ++ suffix
       lazy val coid: Ref.HexString = toBytes.toHexString
       override def toString: String = s"AbsoluteContractId($coid)"
+
+      // We copy how hashcode evaluation is cached in String
+      // http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/lang/String.java
+      private var _hashCode = 0
+
+      override def hashCode(): Int =
+        if (_hashCode != 0)
+          _hashCode
+        else {
+          val hash = MurmurHash3.mixLast(discriminator.hashCode(), suffix.hashCode())
+          _hashCode = if (hash == 0) 1 else hash
+          _hashCode
+        }
+
+      override def equals(obj: Any): Boolean = obj match {
+        case that: V1 =>
+          this.discriminator == that.discriminator && this.suffix == that.suffix
+        case _ => false
+      }
+
     }
 
     object V1 {
+      def apply(discriminator: crypto.Hash): V1 = new V1(discriminator, Bytes.Empty)
+
+      def unapply(arg: V1): Option[(crypto.Hash, Bytes)] =
+        Some((arg.discriminator, arg.suffix))
+
+      def build(discriminator: crypto.Hash, suffix: Bytes): Either[String, V1] =
+        Either.cond(
+          suffix.length <= 94,
+          new V1(discriminator, suffix),
+          s"suffix $suffix to long"
```

Comment:
I recommend to not put the whole suffix into the error message, since we have no idea how long it actually is. How about showing the a length-limited prefix and the number of bytes?

Reply:
fair enough

Reply:
fixed